File: scripts/fetch-hygiene-ratings.py
"""
Fetch Food Hygiene Ratings from the Food Standards Agency (FSA) free public API.
Matches each business by name + postcode and stores the rating.
Run: python scripts/fetch-hygiene-ratings.py

No API key required – FSA API is fully public.
"""

import os
import sys
import time
import json
import re
import psycopg2
import requests
from dotenv import load_dotenv
from difflib import SequenceMatcher

load_dotenv()
load_dotenv(".env")

DB_URL = os.environ.get("DATABASE_URL", "")
FSA_BASE = "https://api.ratings.food.gov.uk"
FSA_HEADERS = {"x-api-version": "2", "Accept": "application/json"}

# ---------------------------------------------------------------------------
# DB helpers
# ---------------------------------------------------------------------------
def get_connection():
    # Use verify-full with system trust store for Neon
    url = DB_URL.replace("sslmode=verify-full", "sslmode=verify-full&sslrootcert=system")
    return psycopg2.connect(url)

def get_businesses(conn):
    with conn.cursor() as cur:
        cur.execute("""
            SELECT id, name, postcode, address
            FROM "Business"
            WHERE postcode IS NOT NULL AND postcode != ''
            ORDER BY name
        """)
        return cur.fetchall()

def update_business(conn, business_id, rating, rating_date):
    with conn.cursor() as cur:
        cur.execute("""
            UPDATE "Business"
            SET "hygieneRating" = %s,
                "hygieneRatingDate" = %s,
                "updatedAt" = NOW()
            WHERE id = %s
        """, (rating, rating_date, business_id))
    conn.commit()

# ---------------------------------------------------------------------------
# FSA API helpers
# ---------------------------------------------------------------------------
def clean_postcode(postcode: str) -> str:
    return postcode.strip().upper().replace(" ", "")

def similarity(a: str, b: str) -> float:
    return SequenceMatcher(None, a.lower(), b.lower()).ratio()

def extract_street_number(name: str) -> str:
    """Strip common business suffixes to improve matching."""
    cleaned = re.sub(r'\b(ltd|limited|plc|co\.?|uk|restaurant|bar|hotel|cafe|coffee|the|&)\b', '', name.lower())
    return cleaned.strip()

def search_fsa(name: str, postcode: str):
    """Search FSA API by postcode first, then name-match."""
    pc = clean_postcode(postcode)
    
    # Strategy 1: search by postcode + partial name
    short_name = name[:20]  # FSA fuzzy matches on partial names
    try:
        resp = requests.get(
            f"{FSA_BASE}/Establishments",
            params={
                "name": short_name,
                "address": pc,
                "pageSize": 10,
                "sortOptionKey": "alpha",
            },
            headers=FSA_HEADERS,
            timeout=10,
        )
        if resp.status_code == 200:
            data = resp.json()
            establishments = data.get("establishments", [])
            if establishments:
                # Pick best name match
                best = max(establishments, key=lambda e: similarity(name, e.get("BusinessName", "")))
                sim = similarity(name, best.get("BusinessName", ""))
                if sim > 0.4:
                    return best

    except Exception as e:
        pass

    # Strategy 2: search by postcode only
    try:
        resp = requests.get(
            f"{FSA_BASE}/Establishments",
            params={
                "address": pc,
                "pageSize": 20,
            },
            headers=FSA_HEADERS,
            timeout=10,
        )
        if resp.status_code == 200:
            data = resp.json()
            establishments = data.get("establishments", [])
            if establishments:
                best = max(establishments, key=lambda e: similarity(name, e.get("BusinessName", "")))
                sim = similarity(name, best.get("BusinessName", ""))
                if sim > 0.45:
                    return best
    except Exception:
        pass

    return None

def parse_rating(establishment: dict) -> tuple:
    """Extract rating value and date from FSA establishment record."""
    rating_value = establishment.get("RatingValue", "")
    rating_date_str = establishment.get("RatingDate", "")
    
    # Normalise rating value
    if rating_value in ["5", "4", "3", "2", "1", "0"]:
        rating = rating_value
    elif rating_value and "Exempt" in str(rating_value):
        rating = "Exempt"
    elif rating_value and "Awaiting" in str(rating_value):
        rating = "AwaitingInspection"
    elif str(rating_value).isdigit():
        rating = str(int(rating_value))
    else:
        rating = None

    # Parse date
    rating_date = None
    if rating_date_str:
        try:
            from datetime import datetime
            # FSA returns dates as "2023-06-01T00:00:00" or "2023-06-01"
            if "T" in rating_date_str:
                rating_date = datetime.fromisoformat(rating_date_str.split("T")[0])
            else:
                rating_date = datetime.fromisoformat(rating_date_str[:10])
        except Exception:
            pass

    return rating, rating_date

# ---------------------------------------------------------------------------
# Progress file
# ---------------------------------------------------------------------------
PROGRESS_FILE = "hygiene-progress.json"

def load_progress():
    if os.path.exists(PROGRESS_FILE):
        with open(PROGRESS_FILE) as f:
            return set(json.load(f))
    return set()

def save_progress(done_ids):
    with open(PROGRESS_FILE, "w") as f:
        json.dump(list(done_ids), f)

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main():
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")

    print("Fetching Food Hygiene Ratings from FSA API...")
    print()

    conn = get_connection()
    businesses = get_businesses(conn)
    total = len(businesses)
    print(f"Found {total} businesses to process\n")

    done_ids = load_progress()
    remaining = [b for b in businesses if b[0] not in done_ids]
    print(f"Resuming: {len(done_ids)} already done, {len(remaining)} remaining\n")

    matched = 0
    not_found = 0
    errors = 0

    for i, (bid, name, postcode, address) in enumerate(remaining, 1):
        try:
            name_safe = name.encode("ascii", "replace").decode()
            est = search_fsa(name, postcode)

            if est:
                rating, rating_date = parse_rating(est)
                if rating is not None:
                    update_business(conn, bid, rating, rating_date)
                    matched += 1
                    fsa_name = est.get("BusinessName", "").encode("ascii", "replace").decode()
                    print(f"  [{i}/{len(remaining)}] {name_safe[:40]:<42} -> {rating} (matched: {fsa_name[:35]})")
                else:
                    not_found += 1
                    print(f"  [{i}/{len(remaining)}] {name_safe[:40]:<42} -> no rating data")
            else:
                not_found += 1
                if i % 20 == 0:
                    print(f"  [{i}/{len(remaining)}] ... {not_found} not found so far")

        except Exception as e:
            errors += 1
            print(f"  ERROR on {name}: {e}")

        done_ids.add(bid)

        # Save progress every 25 businesses
        if i % 25 == 0:
            save_progress(done_ids)
            print(f"    Progress saved ({i}/{len(remaining)})")

        # Be polite to the FSA API - small delay
        time.sleep(0.3)

    save_progress(done_ids)
    conn.close()

    print(f"\n{'='*50}")
    print(f"Complete!")
    print(f"  Matched with hygiene rating: {matched}")
    print(f"  Not found / no rating:       {not_found}")
    print(f"  Errors:                      {errors}")
    print(f"  Total processed:             {matched + not_found + errors}")

if __name__ == "__main__":
    main()

File: scripts/fetch-hygiene-ratings.ts
/**
 * Fetch Food Hygiene Ratings from the FSA free public API.
 * Matches businesses by name + postcode. No API key required.
 * Run: npx tsx scripts/fetch-hygiene-ratings.ts
 */
import "dotenv/config";
import { PrismaClient } from "@prisma/client";
import { PrismaPg } from "@prisma/adapter-pg";
import * as fs from "fs";

const adapter = new PrismaPg({ connectionString: process.env.DATABASE_URL! });
const prisma = new PrismaClient({ adapter });

const FSA_BASE = "https://api.ratings.food.gov.uk";
const FSA_HEADERS = { "x-api-version": "2", Accept: "application/json" };
const PROGRESS_FILE = "hygiene-progress.json";
const DELAY_MS = 350; // polite rate – FSA asks for reasonable use

// ── String similarity (Levenshtein-lite) ──────────────────────────────────
function similarity(a: string, b: string): number {
  const s1 = a.toLowerCase().trim();
  const s2 = b.toLowerCase().trim();
  if (s1 === s2) return 1;
  const longer = s1.length > s2.length ? s1 : s2;
  const shorter = s1.length > s2.length ? s2 : s1;
  if (longer.length === 0) return 1;
  // Count common chars
  let matches = 0;
  const used = new Array(longer.length).fill(false);
  for (let i = 0; i < shorter.length; i++) {
    for (let j = 0; j < longer.length; j++) {
      if (!used[j] && shorter[i] === longer[j]) {
        matches++;
        used[j] = true;
        break;
      }
    }
  }
  return (2 * matches) / (s1.length + s2.length);
}

function cleanPostcode(pc: string): string {
  // FSA API requires standard postcode format WITH space e.g. "PR8 4AR"
  const raw = pc.trim().toUpperCase().replace(/\s+/g, "");
  // Insert space before last 3 chars (standard UK format)
  if (raw.length >= 5) return `${raw.slice(0, -3)} ${raw.slice(-3)}`;
  return raw;
}

function postcodeSector(pc: string): string {
  // e.g. "PR8 4AR" -> "PR8 4", "PR9 0BS" -> "PR9 0"
  const clean = cleanPostcode(pc);
  return clean.split(" ")[0]; // just the outward code, e.g. "PR8"
}

// ── FSA API ───────────────────────────────────────────────────────────────
type FsaEstablishment = {
  FHRSID: number;         // FSA unique establishment ID
  BusinessName: string;
  RatingValue: string;
  RatingDate: string | null;
  AddressLine1?: string;
  PostCode?: string;
};

async function searchFsa(name: string, postcode: string): Promise<FsaEstablishment | null> {
  const pc = cleanPostcode(postcode);
  const outward = postcodeSector(postcode); // e.g. "PR8"

  // Strategy 1: name + full postcode – highest confidence
  try {
    const url = `${FSA_BASE}/Establishments?name=${encodeURIComponent(name.slice(0, 40))}&address=${encodeURIComponent(pc)}&pageSize=5`;
    const res = await fetch(url, { headers: FSA_HEADERS });
    if (res.ok) {
      const data = (await res.json()) as { establishments: FsaEstablishment[] };
      const results = data.establishments || [];
      if (results.length > 0) {
        const best = results.reduce((a, b) =>
          similarity(name, b.BusinessName) > similarity(name, a.BusinessName) ? b : a
        );
        const sim = similarity(name, best.BusinessName);
        const resultOutward = postcodeSector(best.PostCode || "");
        const sameArea = resultOutward === outward || cleanPostcode(best.PostCode || "") === pc;
        if (sim >= 0.55 && sameArea) return best;
        if (sim >= 0.80) return best; // very strong name match, trust regardless
      }
    }
  } catch {/* skip */}

  // Strategy 2: town/city search with name – broader net
  try {
    const url = `${FSA_BASE}/Establishments?name=${encodeURIComponent(name.slice(0, 40))}&address=Southport&pageSize=10`;
    const res = await fetch(url, { headers: FSA_HEADERS });
    if (res.ok) {
      const data = (await res.json()) as { establishments: FsaEstablishment[] };
      // Filter to same outward postcode
      const results = (data.establishments || []).filter((e) =>
        postcodeSector(e.PostCode || "") === outward
      );
      if (results.length > 0) {
        const best = results.reduce((a, b) =>
          similarity(name, b.BusinessName) > similarity(name, a.BusinessName) ? b : a
        );
        if (similarity(name, best.BusinessName) >= 0.60) return best;
      }
    }
  } catch {/* skip */}

  return null;
}

function parseRating(est: FsaEstablishment): { rating: string | null; date: Date | null; fhrsId: string | null } {
  const rv = est.RatingValue?.trim() || "";
  let rating: string | null = null;

  if (["5", "4", "3", "2", "1", "0"].includes(rv)) {
    rating = rv;
  } else if (rv.toLowerCase().includes("exempt")) {
    rating = "Exempt";
  } else if (rv.toLowerCase().includes("awaiting")) {
    rating = "AwaitingInspection";
  } else if (/^\d+$/.test(rv)) {
    rating = rv;
  }

  let date: Date | null = null;
  if (est.RatingDate) {
    try {
      date = new Date(est.RatingDate.split("T")[0]);
      if (isNaN(date.getTime())) date = null;
    } catch {/* skip */}
  }

  const fhrsId = est.FHRSID ? String(est.FHRSID) : null;

  return { rating, date, fhrsId };
}

// ── Progress ──────────────────────────────────────────────────────────────
function loadProgress(): Set<string> {
  if (fs.existsSync(PROGRESS_FILE)) {
    return new Set(JSON.parse(fs.readFileSync(PROGRESS_FILE, "utf-8")));
  }
  return new Set();
}

function saveProgress(done: Set<string>) {
  fs.writeFileSync(PROGRESS_FILE, JSON.stringify([...done]));
}

// ── Main ──────────────────────────────────────────────────────────────────
async function main() {
  console.log("Fetching Food Hygiene Ratings from FSA API...\n");

  const businesses = await prisma.business.findMany({
    where: { postcode: { not: "" } },
    select: { id: true, name: true, postcode: true, address: true, fhrsId: true },
    orderBy: { name: "asc" },
  });

  console.log(`Found ${businesses.length} businesses\n`);

  const done = loadProgress();
  // Process: not yet done AND (no fhrsId saved yet, OR never processed)
  const remaining = businesses.filter((b) => !done.has(b.id) || !b.fhrsId);
  console.log(`Resuming: ${done.size} done previously, ${remaining.length} to process (including fhrsId backfill)\n`);
  console.log(`Resuming: ${done.size} done, ${remaining.length} to process\n`);

  let matched = 0;
  let notFound = 0;

  for (let i = 0; i < remaining.length; i++) {
    const b = remaining[i];
    const label = `[${i + 1}/${remaining.length}]`;

    try {
      const est = await searchFsa(b.name, b.postcode);

      if (est) {
        const { rating, date, fhrsId } = parseRating(est);
        if (rating !== null) {
          await prisma.business.update({
            where: { id: b.id },
            data: {
              hygieneRating: rating,
              hygieneRatingDate: date ?? undefined,
              fhrsId: fhrsId ?? undefined,
            },
          });
          matched++;
          const fsaName = est.BusinessName.slice(0, 35).replace(/[^\x20-\x7E]/g, "?");
          const nameSafe = b.name.slice(0, 38).replace(/[^\x20-\x7E]/g, "?").padEnd(40);
          console.log(`  ${label} ${nameSafe} -> ${rating}  FHRS:${fhrsId}  (${fsaName})`);
        } else {
          notFound++;
        }
      } else {
        notFound++;
        // Only print every 25 not-founds to keep output clean
        if (notFound % 25 === 0) {
          console.log(`  ${label} ... ${notFound} not matched so far`);
        }
      }
    } catch (err) {
      console.error(`  ERROR ${b.name}: ${err}`);
    }

    done.add(b.id);
    if ((i + 1) % 50 === 0) {
      saveProgress(done);
      console.log(`  ── progress saved (${i + 1}/${remaining.length}) matched: ${matched} ──`);
    }

    await new Promise((r) => setTimeout(r, DELAY_MS));
  }

  saveProgress(done);
  await prisma.$disconnect();

  console.log(`\n${"=".repeat(50)}`);
  console.log(`Complete!`);
  console.log(`  Matched with hygiene rating : ${matched}`);
  console.log(`  Not found / no rating data  : ${notFound}`);
  console.log(`  Total processed             : ${matched + notFound}`);
}

main().catch(console.error);
